### PR TITLE
(WIP) test gitlab_ci folding

### PR DESCRIPTION
**Pull request by tylerjw**
This is the begging of me adapting moveit_ci to run on either travis or gitlab-ci.  I expect this to be ignored until it is more featureful and complete.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
More of an observation: I realise quite some work has gone into `moveit_ci`, but seeing the last few PRs I cannot avoid the feeling that we're reinventing the wheel here.

`industrial_ci` has support for quite a few of these features already, and a recent PR actually even copied code from it to add features to `moveit_ci`.

Support for Gitlab, Bitbucket and Github actions is already provided in `industrial_ci`, for both ROS 1 and ROS 2.

Would it perhaps be an idea to take another look at it and see whether we can avoid spending time reimplementing features in `moveit_ci` that other packages already have?

And to clarify: I'm not saying that `industrial_ci` is the answer to everything. There were certainly reasons to create `moveit_ci` back in the day. But things have evolved and perhaps it would be worthwhile to see what the current state of it is.

**Comment by tylerjw** [comment]
@gavanderhoorn you might be right but I'd really like to re-use the code coverage reporting feature I helped add and am now working on a project that is being hosted in gitlab.  

It was difficult enough to get that merged here and @rhaschke was very responsive, it looks like `industrial_ci` has many outstanding PRs that are passing CI and have been approved and are passing CI and are quite old.  I'd just be worried that it would be wasted effort to try to contribute to `industrial_ci`.

**Comment by rhaschke** [comment]
I definitely support a new attempt to merge moveit_ci into industrial_ci. industrial_ci is way more modular and thus facilitates support for other CI providers.
The only thing missing from industrial_ci seems to be the newly added coverage testing.
As a first step, I suggest to try building the MoveIt workspace with industrial_ci.
To avoid building docker images from scratch (which was the main reason to develop our moveit_ci fork), we need to explicitly specify our pre-built docker images via `DOCKER_IMAGE` variable.
Doing this for your gitlab repository would be a nice exercise.

**Comment by gavanderhoorn** [comment]
> @gavanderhoorn you might be right but I'd really like to re-use the code coverage reporting feature I helped add and am now working on a project that is being hosted in gitlab.

Which is why I suggested to take a look and see what is missing.

Adding coverage support is something that would be of interest to the more wider `industrial_ci` user base as well, so I can see this getting added.

> It was difficult enough to get that merged here and @rhaschke was very responsive, it looks like `industrial_ci` has many outstanding PRs that are passing CI that where opened months ago. I'd just be worried that it would be wasted effort to try to contribute to `industrial_ci`.

Almost all of those are 'internal' PRs opened by the maintainers.

What I would suggest is to first open an issue to discuss what you're trying to achieve and get some feedback from the maintainers. I'm sure @ipa-mdl would be willing to talk about this.

Opening PRs out of the blue may be a bit sudden.

**Comment by gavanderhoorn** [comment]
> you might be right but I'd really like to re-use the code coverage reporting feature I helped add and am now working on a project that is being hosted in gitlab.

Commenting on this one particularly: I'm not sure I understand the "but" you add there.

I'm suggesting to integrating your work with something that already has quite a large amount of functionality. I'm not saying you should discard whatever you have now.

If you could explain that "but" there perhaps I could clarify my intentions better.

**Comment by tylerjw** [comment]
What I meant to say is that I have to choose on the GitLab project to either use this or use industrial_ci.  They both have drawbacks, my but meant that I want to use the code coverage feature and therefore working on a PR here to enable GitLab compatibility seemed more be prudent.  Your suggestion is that I should contribute to industrial_ci as the other parth to what I wanted is what else I was responding to.

I did try industrial_ci and have yet to get it to work for my GitLab project.  (This is on a privately hosted GitLab CE, not their cloud solution and therefore has some differences from "normal" GitLab).

**Comment by ipa-mdl** [comment]
> This is on a privately hosted GitLab CE, not their cloud solution and therefore has some differences from "normal" GitLab

Actually, that is how most users (including our group) use it.
So it is "normal", but not properly documented (yet?).

**Comment by tylerjw** [comment]
> To avoid building docker images from scratch (which was the main reason to develop our moveit_ci fork), we need to explicitly specify our pre-built docker images via `DOCKER_IMAGE` variable.
> Doing this for your gitlab repository would be a nice exercise.

@ipa-mdl is this implemented in `industrial_ci`?  Codecov is a nice to have feature but this sounds like something that would greatly decrease CI run times.

**Comment by rhaschke** [comment]
Yes, `industrial_ci` support a custom `DOCKER_IMAGE`.

**Comment by gavanderhoorn** [comment]
I was just going to quote the [documentation](https://github.com/ros-industrial/industrial_ci/blob/legacy/doc/index.rst):

> `DOCKER_IMAGE` (default: not set): Selects a Docker images different from default one. Please note, this disables the handling of `ROS_REPOSITORY_PATH` and `ROS_DISTRO` as ROS needs already to be installed in the image.

And the [Use custom Docker images](https://github.com/ros-industrial/industrial_ci/blob/legacy/doc/index.rst#use-custom-docker-images) section.

**Comment by tylerjw** [comment]
I just discovered `industrial_ci` builds with `catkin_make` which outputs a ton, fills the log, and is therefore unusable.

**Comment by ipa-mdl** [comment]
> I just discovered industrial_ci builds with catkin_make

I am not sure how you came to this conclusion.
The `legacy` branch uses `catkin-tools`. With the `master` branch, you can choose between `catkin-tools` (default for ROS1), `catkin_make`, `catkin_make_insolated` and `colcon` (default for ROS2).

**Comment by tylerjw** [comment]
> > I just discovered industrial_ci builds with catkin_make
> 
> I am not sure how you came to this conclusion.
> The `legacy` branch uses `catkin-tools`. With the `master` branch, you can choose between `catkin-tools` (default for ROS1), `catkin_make`, `catkin_make_insolated` and `colcon` (default for ROS2).

I came to this conclusion by searching through the repo for `catkin build` and `no-stats` the search resulted in a some commit not in a branch.  I then searched for `catkin_make` and found it easily.  I assumed that was why my log files were so full.  You corrected me by pointing to where in the code `catkin_tools` is set to a variable.  You also corrected me that the reason for that is that verbose output is the default (which might work for small projects just fine).  I was mostly just tired of trying to debug it and because this already worked and log folding (my actual motivation for trying to use industrial_ci wasn't implemented) so I gave up.

In short I jumped to conclusions and was wrong because I was confused about how industrial_ci works and what the state of various features and defaults where.  Thank you for helping me debug the use of industrial_ci.  If I need something similar for a ROS2 package I will defiantly consider it.  However the default of verbose debug output seems like it'll be really annoying with large projects or anything that depends on a large built-from-srouce package.

As to the state of this PR, it is NOT WORKING.  It was a very quick attempt to get github folding to work and I haven't yet succeeded.  Hopefully soon I can figure this out.  Here is the reference from gitlab on this feature: https://gitlab.com/gitlab-org/gitlab-foss/merge_requests/28642

**Comment by ipa-mdl** [comment]
> As to the state of this PR, it is NOT WORKING. It was a very quick attempt to get github folding to work and I haven't yet succeeded. 

The foding/section feature of Gitlab is a very recent addition and it does not seem to work properly, at least not on gitlab.com.
They recently refactored the job log system and will make this the default in an upcoming version: https://gitlab.com/gitlab-org/gitlab/merge_requests/20235

**Comment by ipa-mdl** [comment]
Update: I have now a working prototype: https://github.com/ros-industrial/industrial_ci/pull/451

**Comment by tylerjw** [comment]
I'm closing this and will track it in an issue.  The discussion here diverged far from this PR.
